hfile.c

```diff
@@ -138,10 +138,12 @@ hFILE *hfile_init_fixed(size_t struct_size, const char *mode,
     return fp;
 }
 
+static const struct hFILE_backend mem_backend;
+
 void hfile_destroy(hFILE *fp)
 {
     int save = errno;
-    if (fp) free(fp->buffer);
+    if (fp && fp->backend != &mem_backend) free(fp->buffer);
```

Comment:
This extra check means data: hopens now leak memory.  Why are we not freeing here?

Changing this back makes test/hfile pass valgrind leak checks provided we also remove the now redundant ```free(internal_buf)``` calls.

I think it is probably fine to accept that obtaining the internal buffer doesn't equate to now also owning this buffer.   If we wish to have the concept of taking ownership (and hence freeing responsibilities) we'd need an additional flag somewhere.